apps/nova/src/app/[locale]/(dashboard)/(leaderboard)/leaderboard/page.tsx

```diff
@@ -86,19 +87,116 @@ async function fetchLeaderboard(
   };
 
   const limit = 20;
+
+  const supabase = await createClient();
+
+  // Get current user
+  const {
+    data: { user },
+    error: authError,
+  } = await supabase.auth.getUser();
+
+  if (authError || !user?.id || !user?.email) {
+    throw new Error('Auth error or missing user');
+  }
+
   const sbAdmin = await createAdminClient();
 
+  // Check user's role and permissions
+  const { data: userRole, error: roleError } = await sbAdmin
+    .from('nova_roles')
+    .select('*')
+    .eq('email', user.email)
+    .single();
+
+  if (roleError) {
+    throw new Error('Error fetching user role:', roleError);
+  }
+
   // Fetch all challenges for filter options
   const { data: challenges, error: challengesError } = await sbAdmin
     .from('nova_challenges')
-    .select('id, title')
+    .select('*')
     .order('title', { ascending: true });
 
   if (challengesError) {
     console.error('Error fetching challenges:', challengesError.message);
     return defaultData;
   }
 
+  // Fetch all whitelists for the challenges
+  const { data: allWhitelists, error: whitelistsError } = await sbAdmin
+    .from('nova_challenge_whitelisted_emails')
+    .select('*')
+    .in(
+      'challenge_id',
+      challenges.map((challenge) => challenge.id)
+    );
+
+  if (whitelistsError) {
+    throw new Error('Error fetching whitelists: ' + whitelistsError.message);
+  }
```

Comment:
please use `select('*, nova_challenge_whitelisted_emails(*)')` check the logic via database-level instead of handling it like this (very unscalable).